src/Reflection/ReflectionFunctionAbstract.php

```diff
@@ -619,4 +619,17 @@ public function getReturnStatementsAst()
 
         return $visitor->getReturnNodes();
     }
+
+    private function findTypeFromAst($namespace, $locatedSource, $type)
+    {
+        $objectType = (new FindTypeFromAst())->__invoke(
+            $this->reflector->getContextFactory()->createForNamespace(
```

Comment:
> as people generally define a single class per file

This is a reasonable assumption, but the library *must* cater for those with multiple classes.

What I was suggesting is instantiate the `[Cached]ContextFactory` directly in the type finder classes, as it already does (with phpdocumentor's own one). Yes, it'd be a tight coupling, but it already is anyway, so we're not making it worse - i.e. you'd *just* be introducing the performance increase by caching the contexts.

> but we can still benefit from not parsing the AST

I think caching the AST parsing should be dealt with separately; it's distinct from the type resolving.